File: examples/crypto/bls12-381/sync-committee.zig
const std = @import("std");
const crypto = @import("crypto");

/// Ethereum Sync Committee Example
///
/// Demonstrates real-world Ethereum 2.0 sync committee:
/// - 512 validator committee structure
/// - Participation tracking with bitfields
/// - Signature aggregation benefits
/// - Light client verification
pub fn main() !void {
    const stdout = std.io.getStdOut().writer();

    try stdout.print("=== Ethereum Sync Committee Example ===\n\n", .{});

    // Constants
    const sync_size = 512;
    const participation_rate = 0.85; // 85%
    const participants = @as(u32, @intFromFloat(@as(f64, @floatFromInt(sync_size)) * participation_rate));

    // 1. Sync Committee Overview
    try stdout.print("1. Sync Committee Overview\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Purpose: Enable light clients to track beacon chain\n", .{});
    try stdout.print("Size: {d} validators (randomly selected)\n", .{sync_size});
    try stdout.print("Period: 256 epochs (~27 hours)\n", .{});
    try stdout.print("Duty: Sign each beacon block header\n\n", .{});

    // 2. Data Structure
    try stdout.print("2. SyncAggregate Data Structure\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("struct SyncAggregate {{\n", .{});
    try stdout.print("  sync_committee_bits: BitVector[{d}];  // 64 bytes\n", .{sync_size});
    try stdout.print("  sync_committee_signature: Signature;  // 48 bytes\n", .{});
    try stdout.print("}}\n\n", .{});

    try stdout.print("Total size: 112 bytes per block\n", .{});
    try stdout.print("vs individual signatures: {d} bytes (worst case)\n\n", .{sync_size * 48});

    // 3. Participation Tracking
    try stdout.print("3. Participation Tracking\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Bitfield encoding:\n", .{});
    try stdout.print("  - Each bit represents one validator\n", .{});
    try stdout.print("  - 1 = participated, 0 = did not participate\n", .{});
    try stdout.print("  - 512 bits = 64 bytes\n", .{});
    try stdout.print("\nExample participation: {d}/{d} validators ({d:.1}%)\n", .{ participants, sync_size, participation_rate * 100.0 });
    try stdout.print("Bitfield stores which {d} validators signed\n\n", .{participants});

    // 4. Signature Aggregation Process
    try stdout.print("4. Signature Aggregation Process\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Step 1: Each participating validator signs block root\n", .{});
    try stdout.print("  sig_i = privkey_i * H(block_root)\n", .{});
    try stdout.print("  using DST: BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_SYNC_COMMITTEE_\n", .{});

    try stdout.print("\nStep 2: Aggregate participating signatures\n", .{});
    try stdout.print("  aggSig = sig_1 + sig_2 + ... + sig_{d}\n", .{participants});
    try stdout.print("  via G1 point addition\n", .{});

    try stdout.print("\nStep 3: Package SyncAggregate\n", .{});
    try stdout.print("  - Set bitfield bits for participants\n", .{});
    try stdout.print("  - Include aggregated signature\n", .{});
    try stdout.print("  - Include in beacon block\n\n", .{});

    // 5. Light Client Verification
    try stdout.print("5. Light Client Verification\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Light client has:\n", .{});
    try stdout.print("  - Current sync committee public keys ({d} G2 points)\n", .{sync_size});
    try stdout.print("  - Received SyncAggregate (112 bytes)\n", .{});

    try stdout.print("\nVerification steps:\n", .{});
    try stdout.print("  1. Parse bitfield to identify participants\n", .{});
    try stdout.print("  2. Aggregate {d} public keys (G2 addition)\n", .{participants});
    try stdout.print("  3. Hash block root to G1 point\n", .{});
    try stdout.print("  4. Pairing check: e(aggSig, G2) = e(H(block), aggPubKey)\n", .{});
    try stdout.print("  5. Check participation >= 2/3 threshold\n\n", .{});

    // 6. Performance Analysis
    try stdout.print("6. Performance Analysis\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Bandwidth:\n", .{});
    try stdout.print("  - Individual: {d} bytes\n", .{participants * 48});
    try stdout.print("  - Aggregated: 112 bytes\n", .{});
    const compression = @as(f64, @floatFromInt(participants * 48)) / 112.0;
    try stdout.print("  - Compression: {d:.1}x\n", .{compression});

    try stdout.print("\nVerification time (estimated):\n", .{});
    try stdout.print("  - Individual: {d} sigs × 2ms = {d}ms\n", .{ participants, participants * 2 });
    try stdout.print("  - Aggregated: ~2-3ms (2 pairings)\n", .{});
    const speedup = @as(f64, @floatFromInt(participants * 2)) / 2.5;
    try stdout.print("  - Speedup: ~{d:.0}x\n\n", .{speedup});

    // 7. Supermajority Threshold
    try stdout.print("7. Supermajority Threshold\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    const threshold = sync_size * 2 / 3;
    const has_supermajority = participants >= threshold;

    try stdout.print("Required threshold: 2/3 = {d} validators\n", .{threshold});
    try stdout.print("Current participation: {d} validators\n", .{participants});
    try stdout.print("Has supermajority: {}\n", .{has_supermajority});
    try stdout.print("\nLight client decision: {s}\n\n", .{if (has_supermajority) "ACCEPT BLOCK" else "REJECT BLOCK"});

    // 8. Sync Committee Rotation
    try stdout.print("8. Sync Committee Rotation\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    const epochs_per_period = 256;
    const slots_per_epoch = 32;
    const seconds_per_slot = 12;
    const period_seconds = epochs_per_period * slots_per_epoch * seconds_per_slot;
    const period_hours = @as(f64, @floatFromInt(period_seconds)) / 3600.0;

    try stdout.print("Rotation period: {d} epochs\n", .{epochs_per_period});
    try stdout.print("Duration: {d:.1} hours\n", .{period_hours});
    try stdout.print("\nRotation process:\n", .{});
    try stdout.print("  1. New committee randomly selected from active validators\n", .{});
    try stdout.print("  2. Selection based on shuffling algorithm + epoch seed\n", .{});
    try stdout.print("  3. Committee announced 1 period in advance\n", .{});
    try stdout.print("  4. Smooth transition at period boundary\n\n", .{});

    // 9. Security Properties
    try stdout.print("9. Security Properties\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    try stdout.print("Rogue key protection:\n", .{});
    try stdout.print("  - All validators must submit proof-of-possession\n", .{});
    try stdout.print("  - PoP verified before validator activation\n", .{});
    try stdout.print("  - Safe to aggregate without additional checks\n", .{});

    try stdout.print("\nConsensus security:\n", .{});
    try stdout.print("  - 2/3 supermajority required (Byzantine fault tolerant)\n", .{});
    try stdout.print("  - Random committee selection (hard to corrupt)\n", .{});
    try stdout.print("  - Short rotation period (limits attack window)\n\n", .{});

    // 10. Real-World Impact
    try stdout.print("10. Real-World Impact\n", .{});
    try stdout.print("{s}\n", .{"-" ** 60});

    const blocks_per_day = 24 * 60 * 60 / seconds_per_slot;
    const bandwidth_mb = @as(f64, @floatFromInt(blocks_per_day * 112)) / 1024.0 / 1024.0;

    try stdout.print("Blocks per day: {d}\n", .{blocks_per_day});
    try stdout.print("Light client bandwidth: {d:.2} MB/day\n", .{bandwidth_mb});
    try stdout.print("\nEnables:\n", .{});
    try stdout.print("  - Mobile light clients (low bandwidth)\n", .{});
    try stdout.print("  - Browser-based clients (no sync required)\n", .{});
    try stdout.print("  - IoT devices (resource constrained)\n", .{});
    try stdout.print("  - Cross-chain bridges (trustless proofs)\n\n", .{});

    try stdout.print("Without signature aggregation:\n", .{});
    const full_bandwidth_mb = @as(f64, @floatFromInt(blocks_per_day * participants * 48)) / 1024.0 / 1024.0;
    try stdout.print("  - Required bandwidth: {d:.2} MB/day\n", .{full_bandwidth_mb});
    try stdout.print("  - Verification time: {d} hours/day\n", .{@as(f64, @floatFromInt(blocks_per_day * participants * 2)) / 1000.0 / 3600.0});
    try stdout.print("  - Impractical for resource-constrained devices\n\n", .{});

    try stdout.print("=== Complete ===\n", .{});
}

File: examples/crypto/bls12-381/sync-committee.ts
import { bls12_381 } from "@noble/curves/bls12-381.js";
import { randomBytes } from "crypto";

/**
 * Ethereum Sync Committee Example
 *
 * Demonstrates real-world Ethereum 2.0 sync committee:
 * - 512 validator sync committee
 * - Partial participation tracking
 * - Signature aggregation for light clients
 * - Efficient verification
 */

console.log("=== Ethereum Sync Committee Example ===\n");

// Constants
const SYNC_COMMITTEE_SIZE = 512;
const PARTICIPATION_RATE = 0.85; // 85% participation

// Helper: Sign message
function signMessage(privateKey: bigint, message: Uint8Array) {
	const msgHash = bls12_381.G1.hashToCurve(message, {
		DST: "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_SYNC_COMMITTEE_",
	});
	return msgHash.multiply(privateKey);
}

// 1. Setup Sync Committee
console.log("1. Setup Sync Committee");
console.log("-".repeat(60));

// Generate 512 validators (use small indices for demo)
const syncCommittee = Array.from({ length: SYNC_COMMITTEE_SIZE }, (_, i) => {
	// In production, use secure random private keys
	const privKey = BigInt(1000 + i);
	const pubKey = bls12_381.G2.Point.BASE.multiply(privKey);
	return {
		index: i,
		privKey,
		pubKey,
		pubKeyBytes: pubKey.toRawBytes(true), // Compressed 96 bytes
	};
});

console.log("Sync committee size:", SYNC_COMMITTEE_SIZE);
console.log("First validator privkey:", syncCommittee[0].privKey);
console.log(
	"First validator pubkey (compressed):",
	Buffer.from(syncCommittee[0].pubKeyBytes).toString("hex").slice(0, 32),
	"...",
);
console.log("Public key size: 96 bytes compressed\n");

// 2. Beacon Block to Sign
console.log("2. Beacon Block to Sign");
console.log("-".repeat(60));

// Simulate beacon block root (32 bytes)
const blockRoot = randomBytes(32);
const blockRootHex = "0x" + Buffer.from(blockRoot).toString("hex");

console.log("Block root:", blockRootHex.slice(0, 20), "...");
console.log("All sync committee members sign the same block root\n");

// 3. Simulate Participation
console.log("3. Simulate Validator Participation");
console.log("-".repeat(60));

// Random participation (85% of validators online)
const participationBits = Array.from(
	{ length: SYNC_COMMITTEE_SIZE },
	() => Math.random() < PARTICIPATION_RATE,
);
const participantCount = participationBits.filter((p) => p).length;

console.log(
	"Expected participation:",
	(PARTICIPATION_RATE * 100).toFixed(1),
	"%",
);
console.log("Actual participants:", participantCount, "/", SYNC_COMMITTEE_SIZE);
console.log(
	"Participation rate:",
	((participantCount / SYNC_COMMITTEE_SIZE) * 100).toFixed(1),
	"%",
);

// Encode participation as bitfield (64 bytes for 512 bits)
const participationBytes = new Uint8Array(64);
for (let i = 0; i < SYNC_COMMITTEE_SIZE; i++) {
	if (participationBits[i]) {
		const byteIndex = Math.floor(i / 8);
		const bitIndex = i % 8;
		participationBytes[byteIndex] |= 1 << bitIndex;
	}
}

console.log("Participation bitfield: 64 bytes");
console.log(
	"Bitfield (first 16 bytes):",
	Buffer.from(participationBytes.subarray(0, 16)).toString("hex"),
	"...\n",
);

// 4. Collect Signatures
console.log("4. Collect Signatures from Participants");
console.log("-".repeat(60));

const participatingValidators = syncCommittee.filter(
	(_, i) => participationBits[i],
);
const signatures = participatingValidators.map((v) =>
	signMessage(v.privKey, blockRoot),
);

console.log("Signatures collected:", signatures.length);
console.log(
	"First signature (x):",
	signatures[0].toAffine().x.toString(16).slice(0, 16),
	"...",
);
console.log(
	"Last signature (x):",
	signatures[signatures.length - 1].toAffine().x.toString(16).slice(0, 16),
	"...\n",
);

// 5. Aggregate Signatures
console.log("5. Aggregate Signatures");
console.log("-".repeat(60));

const startAgg = Date.now();
let aggregatedSignature = signatures[0];
for (let i = 1; i < signatures.length; i++) {
	aggregatedSignature = aggregatedSignature.add(signatures[i]);
}
const aggTime = Date.now() - startAgg;

console.log(
	"Individual signatures:",
	signatures.length,
	"× 48 bytes =",
	signatures.length * 48,
	"bytes",
);
console.log("Aggregated signature: 1 × 48 bytes = 48 bytes");
console.log(
	"Compression ratio:",
	((signatures.length * 48) / 48).toFixed(1) + "x",
);
console.log("Aggregation time:", aggTime, "ms");
console.log(
	"\nAggregated signature (x):",
	aggregatedSignature.toAffine().x.toString(16).slice(0, 16),
	"...\n",
);

// 6. SyncAggregate Structure
console.log("6. SyncAggregate Data Structure");
console.log("-".repeat(60));

interface SyncAggregate {
	syncCommitteeBits: Uint8Array; // 64 bytes (512 bits)
	syncCommitteeSignature: Uint8Array; // 48 bytes compressed
}

const syncAggregate: SyncAggregate = {
	syncCommitteeBits: participationBytes,
	syncCommitteeSignature: aggregatedSignature.toRawBytes(true),
};

const totalSize =
	syncAggregate.syncCommitteeBits.length +
	syncAggregate.syncCommitteeSignature.length;
console.log("SyncAggregate structure:");
console.log("  - sync_committee_bits: 64 bytes (bitfield)");
console.log("  - sync_committee_signature: 48 bytes (aggregated)");
console.log("  - Total size:", totalSize, "bytes");
console.log("\nvs individual signatures:", participantCount * 48, "bytes");
console.log(
	"Bandwidth savings:",
	(1 - totalSize / (participantCount * 48)).toFixed(2) + "x or",
	((1 - totalSize / (participantCount * 48)) * 100).toFixed(1) + "%\n",
);

// 7. Verification (Light Client)
console.log("7. Verification by Light Client");
console.log("-".repeat(60));

// Light client has sync committee public keys
const participatingPubKeys = participatingValidators.map((v) => v.pubKey);

// Aggregate public keys
let aggregatedPubKey = participatingPubKeys[0];
for (let i = 1; i < participatingPubKeys.length; i++) {
	aggregatedPubKey = aggregatedPubKey.add(participatingPubKeys[i]);
}

// Verify aggregated signature
const messageHash = bls12_381.G1.hashToCurve(blockRoot, {
	DST: "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_POP_SYNC_COMMITTEE_",
});

const startVerify = Date.now();
const lhs = bls12_381.pairing(aggregatedSignature, bls12_381.G2.Point.BASE);
const rhs = bls12_381.pairing(messageHash, aggregatedPubKey);
const valid = lhs.equals(rhs);
const verifyTime = Date.now() - startVerify;

console.log("Light client verification:");
console.log("  1. Parse participation bitfield");
console.log("  2. Aggregate", participantCount, "public keys");
console.log("  3. Hash block root to G1 point");
console.log("  4. Single pairing check");
console.log("\nVerification result:", valid ? "VALID" : "INVALID");
console.log("Verification time:", verifyTime, "ms");
console.log(
	"\nvs individual verification:",
	participantCount,
	"× ~2ms ≈",
	participantCount * 2,
	"ms",
);
console.log(
	"Speedup:",
	Math.round((participantCount * 2) / verifyTime) + "x faster\n",
);

// 8. Supermajority Check
console.log("8. Supermajority Threshold");
console.log("-".repeat(60));

const SUPERMAJORITY_THRESHOLD = Math.floor((SYNC_COMMITTEE_SIZE * 2) / 3); // 66.7%
const hasSupermajority = participantCount >= SUPERMAJORITY_THRESHOLD;

console.log(
	"Supermajority threshold: 2/3 =",
	SUPERMAJORITY_THRESHOLD,
	"validators",
);
console.log("Actual participants:", participantCount);
console.log("Has supermajority:", hasSupermajority);
console.log(
	"\nLight client accepts block:",
	hasSupermajority && valid ? "YES" : "NO\n",
);

// 9. Rotation (Sync Committee Updates)
console.log("9. Sync Committee Rotation");
console.log("-".repeat(60));

const EPOCHS_PER_SYNC_COMMITTEE_PERIOD = 256; // 256 epochs ≈ 27 hours
const SLOTS_PER_EPOCH = 32;
const SECONDS_PER_SLOT = 12;

const periodDuration =
	EPOCHS_PER_SYNC_COMMITTEE_PERIOD * SLOTS_PER_EPOCH * SECONDS_PER_SLOT;
const hours = periodDuration / 3600;

console.log(
	"Sync committee period:",
	EPOCHS_PER_SYNC_COMMITTEE_PERIOD,
	"epochs",
);
console.log("Duration:", hours.toFixed(1), "hours");
console.log("\nRotation process:");
console.log("  1. New committee randomly selected");
console.log("  2. Current committee signs blocks");
console.log("  3. Next committee prepared in advance");
console.log("  4. Smooth transition at period boundary\n");

// 10. Real-World Performance
console.log("10. Real-World Performance Implications");
console.log("-".repeat(60));

const blocksPerDay = (24 * 60 * 60) / SECONDS_PER_SLOT;
const syncAggregatePerDay = blocksPerDay;
const bandwidthPerDay = syncAggregatePerDay * totalSize;

console.log("Blocks per day:", blocksPerDay);
console.log("SyncAggregate per day:", syncAggregatePerDay);
console.log(
	"Bandwidth per day:",
	(bandwidthPerDay / 1024 / 1024).toFixed(2),
	"MB",
);
console.log(
	"\nvs individual signatures per day:",
	((syncAggregatePerDay * participantCount * 48) / 1024 / 1024).toFixed(2),
	"MB",
);
console.log(
	"Daily bandwidth savings:",
	(
		(syncAggregatePerDay * participantCount * 48 - bandwidthPerDay) /
		1024 /
		1024
	).toFixed(2),
	"MB\n",
);

console.log("Light client benefits:");
console.log("  - Minimal bandwidth (112 bytes per block)");
console.log("  - Fast verification (single pairing)");
console.log("  - High security (supermajority consensus)");
console.log("  - Enables mobile/browser light clients\n");

console.log("=== Complete ===");
